### PR TITLE
Don't update dependencies more than once

### LINKED ISSUE #120: Corral update error
I get a lot of errors if I use the following corral.json and run `corral update` before running `corral fetch`. 

```
{
  "deps": [
    {
      "locator": "github.com/niclash/jay.git"
    },
    {
      "locator": "github.com/ponylang/crypto.git"
    },
    {
      "locator": "github.com/niclash/pony-metric.git"
    },
    {
      "locator": "github.com/niclash/jennet.git"
    },
    {
      "locator": "github.com/niclash/pony-websocket.git"
    },
    {
      "locator": "github.com/ponylang/http.git"
    }
  ],
  "info": {}
}
```

Some of the errors:

```
git checking out @1.0.0 into /home/sean/tmp/ctest/_corral/github_com_ponylang_net_ssl
Error loading dep bundle: github_com_ponylang_regex
Error loading dep bundle: github_com_ponylang_regex
  exit: Exited(128)
  out:
  err: fatal: Unable to create '/home/sean/tmp/ctest/_repos/github_com_ponylang_regex_git/.git/index.lock': File exists.

Another git process seems to be running in this repository, e.g.
an editor opened by 'git commit'. Please make sure all processes
are terminated then try again. If it still fails, a git process
may have crashed in this repository earlier:
remove the file manually to continue.

git checking out @1.0.0 into /home/sean/tmp/ctest/_corral/github_com_ponylang_regex
git checking out @1.0.0 into /home/sean/tmp/ctest/_corral/github_com_ponylang_regex
git checking out @1.0.0 into /home/sean/tmp/ctest/_corral/github_com_ponylang_net_ssl
Error loading dep bundle: github_com_ponylang_net_ssl
Error loading dep bundle: github_com_ponylang_net_ssl
  exit: Exited(128)
  out:
  err: error: unable to create file /home/sean/tmp/ctest/_corral/github_com_ponylang_net_ssl/assets/cert.pem: File exists
```

### PR DESCRIPTION
**Pull request by SeanTAllen**
Prior to this commit, when update would run, if there were transitive dependencies, it might try to fetch/update them more than once potentially leading to errors.

This commit adds a regression unit test that shows the problem and a simple fix to the _Updater to handle. I use "simple" to mean "small" or "baby-steps". The internal design of _Updater needs to be examined in more depth in order to make this all better.

I'm not tackling that now. Before getting into design, we need to add more tests in general to see what affordances for insight and tracing that corral needs to provide then we can build the design around that.

I wanted to add tests to demonstrate the bug and then show that the
bug was fixed by my changes. This was problematic because there are no real unit tests for any
commands.

This PR includes a first step in providing better unit testability
of various parts of corral in general and commands specifically.

There are a number of testing-related changes that are brought this PR.
As we move forward with making corral more testable, some of these changes
will probably fall away. I did what I considered the minimal thing to get
a unit-test harness for testing _Updater working. 

The approach I have taken does leave a lot of things in an inconsistent state.
For example, this PR introduces a new way of organizing unit tests that keeps
said tests near the code they are testing and allows for testing private implementations
like the _Update that is part of the update cmd.

These inconsistencies are temporary and as we move forward with adding more tests, will
start to fall away.

Closes #120

### CODE REVIEW

**Comment by SeanTAllen** [comment]
This PR needs to be squashed and the top level GH comment should be used as the body of the squash and the PR title should be the commit title.

File: corral/cmd/_test_cmd_update.pony [hunk 1]
```diff
@@ -37,6 +39,44 @@ class iso _TestEmptyDeps is UnitTest
...
+    let log = Log(LvlNone, h.env.err, SimpleLogFormatter)
+    let fp: FilePath = _TestData.file_path_from(h, "regression-120/bundle-entrypoint")?
+    let repo_cache = _TestRepoCache(auth)?
+    let ctx = Context(h.env, log, log, false, repo_cache)
+    let project = Project(auth, log, fp)
+    let bundle = Bundle.load(fp, log)?
+    let recorder = _OpsRecorder(h, 4, 4, 4)
+    let vcs_builder: VCSBuilder = _TestCmdUpdateVCSBuilder(recorder)
+
+    // when
```
Comment (Theodus) [inline corral/cmd/_test_cmd_update.pony:72]: Is this comment intended to be here?

Comment (SeanTAllen) [reply corral/cmd/_test_cmd_update.pony:72]: Yes...

There's also a "given" above before the setup so Given this setup, when Blah...

from a given, when, then style.

If you think it is confusing, I can remove.

Comment (Theodus) [reply corral/cmd/_test_cmd_update.pony:72]: I think it can stay, since I don't think it would hinder anyone's understanding of the test

**Pull request merged**